component/src/main/java/com/celements/photo/utilities/DefaultImageUrlExtractor.java

```diff
@@ -0,0 +1,119 @@
+
+import javax.inject.Singleton;
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xwiki.component.annotation.Component;
+
+import com.celements.model.context.ModelContext;
+import com.celements.photo.container.ImageUrlDim;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.web.Utils;
+
+import groovy.lang.Immutable;
+
+@Immutable
+@Singleton
+@Component
+public class DefaultImageUrlExtractor implements ImageUrlExtractor {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(DefaultImageUrlExtractor.class);
+
+  public static final int MAX_ALLOWED_DIM = 1000000000;
+  public static final int MIN_SOCIAL_MEDIA_IMAGE_SIZE = 200;
+  public static final int MIN_SOCIAL_MEDIA_AREA_SIZE = MIN_SOCIAL_MEDIA_IMAGE_SIZE
+      * MIN_SOCIAL_MEDIA_IMAGE_SIZE;
+
+  private final Pattern IMG_FROM_HTML_PATTERN = Pattern.compile("<img .*?src=['\"](.*?)['\"].*?/>");
+  private final Pattern EXTRACT_ACTION_PATTERN = Pattern.compile("^.*?/(.*?)/.*$");
+  private final Pattern EXTRACT_SPACE_PATTERN = Pattern.compile("^(.*?/){2}(.*?)/.*$");
+  private final Pattern EXTRACT_DOCUMENT_PATTERN = Pattern.compile("^(.*?/){3}(.*?)/.*$");
+  private final Pattern EXTRACT_FILENAME_PATTERN = Pattern.compile("^(.*?/){4}(.*?)(|\\?.*)$");
+
+  @Override
+  public @NotNull List<ImageUrlDim> extractImagesList(@NotNull String content) {
+    Map<Long, List<ImageUrlDim>> articleImages = new TreeMap<>();
+    Matcher m = IMG_FROM_HTML_PATTERN.matcher(content);
+    while (m.find()) {
+      String imgUrl = m.group(1);
+      Long key = getImgUrlSizeKey(imgUrl);
+      if (!articleImages.containsKey(key)) {
+        articleImages.put(key, new ArrayList<ImageUrlDim>());
+      }
+      articleImages.get(key).add(getImgUrlExternal(imgUrl));
+    }
+    return filterImagesByMissingOrSmallSize(articleImages);
+  }
+
+  List<ImageUrlDim> filterImagesByMissingOrSmallSize(Map<Long, List<ImageUrlDim>> articleImages) {
+    List<ImageUrlDim> sortedImages = new ArrayList<>();
+    for (Long imgArea : articleImages.keySet()) {
+      // Image size not extractable is in key == '-1' Don't include too small images.
+      if ((imgArea == -1) || (imgArea >= (MIN_SOCIAL_MEDIA_AREA_SIZE))) {
+        Collections.reverse(articleImages.get(imgArea));
+        sortedImages.addAll(articleImages.get(imgArea));
+      }
+    }
+    return sortedImages;
+  }
+
+  ImageUrlDim getImgUrlExternal(String imgUrl) {
+    if (!imgUrl.startsWith("http://") && !imgUrl.startsWith("https://")) {
+      String action = getMatchedPart(EXTRACT_ACTION_PATTERN.matcher(imgUrl), 1);
+      String space = getMatchedPart(EXTRACT_SPACE_PATTERN.matcher(imgUrl), 2);
+      String docname = getMatchedPart(EXTRACT_DOCUMENT_PATTERN.matcher(imgUrl), 2);
+      String filename = getMatchedPart(EXTRACT_FILENAME_PATTERN.matcher(imgUrl), 2);
+      String query = imgUrl.substring(Math.min(imgUrl.indexOf('?') + 1, imgUrl.length()));
```

Comment:
imgUrl.length() not needed it is default behaviour of substring please remove.
https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#substring(int)